basilica/basilica/__init__.py

```diff
@@ -81,16 +86,60 @@ def raw_embed(self, url, data, opts, timeout):
     # TODO: parallelize
     def embed(self, url, data, batch_size, opts, timeout):
         batch = []
+        outstanding = 0
+        api_thread = threading.Thread(target=self.raw_embed_wrapper, args=(url, opts, timeout))
+        api_thread.setDaemon(True)
+        api_thread.start()
         for i in data:
             batch.append(i)
             if len(batch) >= batch_size:
-                for e in self.raw_embed(url, batch, opts=opts, timeout=timeout):
-                    yield e
+                self.write_queue.put_nowait(batch)
                 batch = []
+                if self.thread_running:
+                    emb = self.read_queue.get(block=True)
+                    if isinstance(emb, Exception):
+                        raise emb
+                    for e in emb:
+                        yield e
+                try:
+                    emb = self.read_queue.get(block=False) 
+                except Queue.Empty:
+                    continue
+                else:
+                    for e in emb:
+                        yield e
         if len(batch) > 0:
-            for e in self.raw_embed(url, batch, opts=opts, timeout=timeout):
-                yield e
-            batch = []
+            self.write_queue.put_nowait(batch)
+        while not self.write_queue.empty() or self.thread_running:
+            try:
+                emb = self.read_queue.get(block=True)
+                if isinstance(emb, Exception):
+                    raise emb
+            except Queue.Empty:
+                continue
+            except Exception as e:
+                raise e
+            else:
+                for e in emb:
+                    yield e
+
+    def raw_embed_wrapper(self, url, opts, timeout):
+        while True:
+            try:
+                batch = self.write_queue.get(block=False)
+                self.thread_running = True
+                emb = self.raw_embed(url, batch, opts=opts, timeout=timeout)
+                self.read_queue.put(emb, block=False)
+                self.thread_running = False if self.write_queue.empty() else True
+            except Queue.Empty:
+                continue
+            except Exception as e:
+                self.read_queue.put(e, block=True)
+                self.thread_running = False
+                break
+            if self.thread_running is False and self.write_queue.empty():
+                break
```

Comment:
I think as written this code will exit prematurely if the API calls return faster than we can queue up more data.

Reply:
We should probably have a test where the iterator returning the sentences to be embedded is very slow so that we can catch mistakes like this.

Reply:
Probably the way to fix this is to have `embed` put a "we're done" message on the queue, and have the worker exit when it receives it.